src/util/basic_types.hpp

```diff
@@ -41,9 +95,103 @@ template<> struct truncate_check_impl<int64_t, size_t, 0> {
   }
 };
 
+// Convenience macro for declaring types allocated via shared_ptr.
+#define DECL_STRUCT(X) \
+  struct X; \
+  using X ## _p = shared_ptr<X>; \
+  bool operator==(shared_ptr<X>, shared_ptr<X>); \
+
 // For integer values, returns ceil(n / m);
 template<typename T> T ceil_divide(T n, T m) {
   return (n + (m-1)) / m;
 }
 
+using i8 = int8_t;
```

Comment:
Not sure I like these typedefs; now we have two ways of referring to each of these types within our code. Let's stick to one or the other throughout? (I'd be fine with either dropping the typedefs, or a second PR that converts all uses throughout turicreate of `[s]int[n]_t` to the typedefs. But let's have a plan for how to reconcile the two uses.)